clients/drcachesim/reader/config_reader.cpp

```diff
@@ -0,0 +1,320 @@
+ * DAMAGE.
+ */
+
+#include "config_reader.h"
+
+#include <iostream>
+using namespace std;
+
+config_reader_t::config_reader_t()
+{
+    /* Empty. */
+}
+
+config_reader_t::~config_reader_t()
+{
+    fin.close();
+}
+
+bool
+config_reader_t::configure(const string &config_file,
+                           cache_simulator_knobs_t &knobs,
+                           std::map<string, cache_params_t*> &caches)
+{
+    // Open the config file.
+    fin.open(config_file);
+    if (!fin.is_open()) {
+        ERRMSG("Failed to open the config file '%s'\n", config_file.c_str());
+        return false;
+    }
+
+    // Walk through the configuration file.
+    while (!fin.eof()) {
+        string param;
+        fin >> ws >> param;
+
+        if (param == "//") {
+            // A comment.
+            getline(fin, param);
+        }
+        else if (param == "num_cores") {
+            // Number of cache cores.
+            fin >> knobs.num_cores;
+            if (knobs.num_cores == 0) {
+                ERRMSG("Number of cores must be >0\n");
+                return false;
+            }
+        }
+        else if (param == "line_size") {
+            // Cache line size in bytes.
+            fin >> knobs.line_size;
+            if (knobs.line_size == 0) {
+                ERRMSG("Line size must be >0\n");
+                return false;
+            }
+        }
+        else if (param == "skip_refs") {
+            // Number of references to skip.
+            fin >> knobs.skip_refs;
+        }
+        else if (param == "warmup_refs") {
+            // Number of references to use for caches warmup.
+            fin >> knobs.warmup_refs;
+        }
+        else if (param == "warmup_fraction") {
+            // Fraction of cache lines that must be filled to end the warmup.
+            fin >> knobs.warmup_fraction;
+            if (knobs.warmup_fraction < 0.0 || knobs.warmup_fraction > 1.0) {
+                ERRMSG("Warmup fraction should be in [0.0, 1.0]\n");
+                return false;
+            }
+        }
+        else if (param == "sim_refs") {
+            // Number of references to simulate.
+            fin >> knobs.sim_refs;
+        }
+        else if (param == "cpu_scheduling") {
+            // Whether to simulate CPU scheduling or not.
+            string bool_val;
+            fin >> bool_val;
+            if (bool_val == "true" || bool_val == "True" ||
+                bool_val == "TRUE") {
+                knobs.cpu_scheduling = true;
+            }
+            else {
+                knobs.cpu_scheduling = false;
+            }
+        }
+        else if (param == "verbose") {
+            // Verbose level.
+            fin >> knobs.verbose;
+        }
+        else {
+            // A cache unit.
+            cache_params_t *cache = new cache_params_t;
+            cache->name = param;
+            if (!configure_cache(cache)) {
+                return false;
```

Comment:
delete cache

Reply:
?  I don't see it being returned to the caller on this "return false" path where my comment is.

Reply:
I am wondering if there is a strong reason to use dynamic allocations here at all, or could we use a map of objects, allocate the cache objects on the stack and move them into the map here?

Reply:
Derek, I misunderstood your initial comment. I thought you wanted to delete the object even if the call succeeds. I am deleting it if the call fails as you suggested.

Florian, in whose stack frame should the objects be allocated? the caller? how many objects? we do not know beforehand how many objects are in the config file.

Reply:
If we use `std::map<string, cache_params_t>` we can create the objects on the stack here and move it into the map. The map, created in the caller, will own the objects.

Reply:
+1, this will make cleanup for the caller easier as well

Reply:
Done.